From bbeafc850181cc0bf9724c71f1aa0015acdb7480 Mon Sep 17 00:00:00 2001
From: Keith Packard <keithp@keithp.com>
Date: Sat, 11 Oct 2008 21:43:58 -0700
Subject: [PATCH 4/4] Ensure that _XReadEvents always leaves an event in the queue on return

XNextEvent assumes that the event queue will be non-empty on return from
_XReadEvents, but with multiple event readers running, the previous change
could leave the queue empty on return from process_responses. Re-invoke
process_responses until the queue is non-empty.

Signed-off-by: Keith Packard <keithp@keithp.com>
---
 src/xcb_io.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/src/xcb_io.c b/src/xcb_io.c
index fc8e53f..2256a3f 100644
--- a/src/xcb_io.c
+++ b/src/xcb_io.c
@@ -267,7 +267,9 @@ void _XReadEvents(Display *dpy)
 	if(dpy->xcb->event_owner != XlibOwnsEventQueue)
 		return;
 	check_internal_connections(dpy);
-	process_responses(dpy, 1, 0, 0);
+	do {
+		process_responses(dpy, 1, 0, 0);
+	} while (dpy->qlen == 0);
 }
 
 /*
-- 
1.6.0.2.307.gc427.dirty

